### PR TITLE
fix: Link preview should not enlarge small images

### LINKED ISSUE #5509: Link preview should not enlarge small images
### Description:

When a user creates a message containing a link to an image, the app shows a preview of that image. The preview is resized in order for its width to match the width available for the message content. While this works fine for large images, (very) small images get enlarged by enormous factors.

### Environment Information:

- Rocket.Chat Server Version: 6.5.3
- Rocket.Chat App Version: 4.45.0.55377
- Device Name: ELE-L29 (Huawei P30)
- OS Version: EMUI 12.0.0 (Android 11)

### Steps to reproduce:

1. Write a message containing a link to a very small image, e.g. https://rueckgr.at/~paulchen/avatar-small.png (24x24 px).

### Expected behavior:

The image is not unnecessarily enlarged for the link preview. This is how it looks like in a desktop browser:

image

And in the mobile browser:

image

### Actual behavior:

image

### PR DESCRIPTION
**Pull request by JASIM0021**
## Proposed changes

This pull request addresses an issue with image previews being resized disproportionately in the message content. The changes ensure that images are displayed in their original dimensions, preserving their aspect ratios and preventing unnecessary enlargement or shrinkage.

The solution includes:
- Fetching the image's original dimensions using `Image.getSize`.
- Dynamically setting the width and height of the image based on its original dimensions.
- Managing the loading state of the image to handle errors and display the image correctly once loaded.

## Issue(s)

Fixes #5509 

## How to test or reproduce

1. Create a message containing a link to an image, particularly a very small one.
2. Observe the image preview in the message content.

**Expected Result:** 
The image should display in its original size without being disproportionately enlarged.

## Screenshots

Before:
Screenshort_1234

After:

Screenshot_1720886956

## Types of changes

- [x] Bugfix (non-breaking change which fixes an issue)

### CODE REVIEW

File: app/containers/message/Urls.tsx [hunk 1]
```diff
@@ -110,8 +112,18 @@ const Url = React.memo(
 		const hasContent = url.title || url.description;
 
 		let image = url.image || url.url;
+		
+		
+		
+
```
Comment (diegolmello) [inline app/containers/message/Urls.tsx:118]:
```suggestion
		
```

File: app/containers/message/Urls.tsx [hunk 2]
```diff
@@ -110,8 +112,18 @@ const Url = React.memo(
...
 		if (image) {
 			image = image.includes('http') ? image : `${baseUrl}/${image}?rc_uid=${user.id}&rc_token=${user.token}`;
+			Image.getSize(image, (width, height) => {
```
Comment (diegolmello) [inline app/containers/message/Urls.tsx:121]: You're doing this on every render.
Can you make it run once after image is downloaded?

Comment (JASIM0021) [reply app/containers/message/Urls.tsx:118]: In the current implementation, the image dimensions are fetched and set once the image loads. This process triggers a state update which re-renders the component to apply the new dimensions. Here’s a step-by-step outline of what happens:

Initial Render: The component initially renders with imageLoadedState set to 'loading' and imageDimensions set to { width: 0, height: 0 }.
Image Loading: If an image URL is available, the Image.getSize method is called to fetch the image dimensions.
State Update: When the dimensions are fetched successfully, setImageDimensions and setImageLoadedState are called to update the state.
Re-render: The state update triggers a re-render of the component with the new dimensions.

Comment (diegolmello) [reply app/containers/message/Urls.tsx]: Add `console.count('PREVIEW RENDER')` on your code.
You'll see the same logs from the image.

Screenshot 2024-07-23 at 1 34 14 PM

What I asked you to do is what I did on https://github.com/RocketChat/Rocket.Chat.ReactNative/pull/5789/commits/0324cb4490ecbb78074a9ad3f81db0335f61649b

Here's the difference.
image

**Review by diegolmello — APPROVED** [review]

**Comment by diegolmello** [comment]
@JASIM0021 Merged. Thanks for your contribution!

**Pull request merged**